### PR TITLE
Codechange: Remove alloca usages

### PR DESCRIPTION
**Pull request by LordAro**
## Motivation / Problem

`alloca` is a real sketchy function from C that is essentially "just find some memory in the stack frame, and hope"

## Description

Removes all usages of alloca and uses proper C++ functions instead. Mostly std::string & std::vector

## Limitations

Limited testing. Probably off-by-one errors somewhere.

## Checklist for review

Some things are not automated, and forgotten often. This list is a reminder for the reviewers.
* The bug fix is important enough to be backported? (label: 'backport requested')
* This PR touches english.txt or translations? Check the [guidelines](https://github.com/OpenTTD/OpenTTD/blob/master/docs/eints.md)
* This PR affects the save game format? (label 'savegame upgrade')
* This PR affects the GS/AI API? (label 'needs review: Script API')
    * ai_changelog.hpp, gs_changelog.hpp need updating.
    * The compatibility wrappers (compat_*.nut) need updating.
* This PR affects the NewGRF API? (label 'needs review: NewGRF')
    * newgrf_debug_data.h may need updating.
    * [PR must be added to API tracker](https://wiki.openttd.org/en/Development/NewGRF/Specification%20Status)

### CODE REVIEW

File: src/os/windows/win32.cpp [hunk 1]
```diff
@@ -669,13 +669,13 @@ int OTTDStringCompare(const char *s1, const char *s2)
 		int len_s2 = MultiByteToWideChar(CP_UTF8, 0, s2, -1, nullptr, 0);
 
 		if (len_s1 != 0 && len_s2 != 0) {
-			LPWSTR str_s1 = AllocaM(WCHAR, len_s1);
-			LPWSTR str_s2 = AllocaM(WCHAR, len_s2);
+			std::wstring str_s1(len_s1);
+			std::wstring str_s2(len_s2);
```
Comment (glx22) [inline src/os/windows/win32.cpp]:
```suggestion
			std::wstring str_s1(len_s1 + 1, L'\0');
			std::wstring str_s2(len_s2 + 1, L'\0');
```
Went for the same solution as line 547 (including the extra wchar for safety)

File: src/video/win32_v.cpp [hunk 2]
```diff
@@ -1056,8 +1056,7 @@ bool VideoDriver_Win32GDI::AllocateBackingStore(int w, int h, bool force)
 
 	if (!force && w == _screen.width && h == _screen.height) return false;
 
-	BITMAPINFO *bi = (BITMAPINFO *)alloca(sizeof(BITMAPINFOHEADER) + sizeof(RGBQUAD) * 256);
-	memset(bi, 0, sizeof(BITMAPINFOHEADER) + sizeof(RGBQUAD) * 256);
+	BITMAPINFO *bi = (BITMAPINFO *)new char[sizeof(BITMAPINFOHEADER) + sizeof(RGBQUAD * 256)]();
```
Comment (glx22) [inline src/video/win32_v.cpp]:
```suggestion
	BITMAPINFO *bi = (BITMAPINFO *)new char[sizeof(BITMAPINFOHEADER) + sizeof(RGBQUAD) * 256]();
```

File: src/video/win32_v.cpp [hunk 3]
```diff
@@ -346,7 +346,7 @@ static LRESULT HandleIMEComposition(HWND hwnd, WPARAM wParam, LPARAM lParam)
 		if ((lParam & GCS_COMPSTR) && DrawIMECompositionString()) {
 			/* Read composition string from the IME. */
 			LONG len = ImmGetCompositionString(hIMC, GCS_COMPSTR, nullptr, 0); // Length is always in bytes, even in UNICODE build.
-			wchar_t *str = (wchar_t *)_alloca(len + sizeof(wchar_t));
+			std::wstring str(len, '\0');
 			len = ImmGetCompositionString(hIMC, GCS_COMPSTR, str, len);
```
Comment (glx22) [inline src/video/win32_v.cpp]:
```suggestion
			len = ImmGetCompositionString(hIMC, GCS_COMPSTR, str.data(), len);
```
and missing `.c_str()` on lines 355 and 360 (comments not available for theses, so can't make suggestion).

**Review by glx22 — DISMISSED** [review]
Changes I needed to build with VS.

File: src/video/win32_v.cpp [hunk 4]
```diff
@@ -346,7 +346,7 @@ static LRESULT HandleIMEComposition(HWND hwnd, WPARAM wParam, LPARAM lParam)
 		if ((lParam & GCS_COMPSTR) && DrawIMECompositionString()) {
 			/* Read composition string from the IME. */
 			LONG len = ImmGetCompositionString(hIMC, GCS_COMPSTR, nullptr, 0); // Length is always in bytes, even in UNICODE build.
-			wchar_t *str = (wchar_t *)_alloca(len + sizeof(wchar_t));
+			std::wstring str(len, '\0');
```
Comment (glx22) [inline src/video/win32_v.cpp]:
```suggestion
			std::wstring str(len + 1, '\0');
```

File: src/video/win32_v.cpp [hunk 5]
```diff
@@ -328,8 +328,8 @@ static LRESULT HandleIMEComposition(HWND hwnd, WPARAM wParam, LPARAM lParam)
 		if (lParam & GCS_RESULTSTR) {
 			/* Read result string from the IME. */
 			LONG len = ImmGetCompositionString(hIMC, GCS_RESULTSTR, nullptr, 0); // Length is always in bytes, even in UNICODE build.
-			wchar_t *str = (wchar_t *)_alloca(len + sizeof(wchar_t));
-			len = ImmGetCompositionString(hIMC, GCS_RESULTSTR, str, len);
+			std::wstring str(len, '\0');
```
Comment (glx22) [inline src/video/win32_v.cpp]:
```suggestion
			std::wstring str(len + 1, '\0');
```

**Review by glx22 — COMMENTED** [review]
And of course I noticed the obiwan after the first review.

File: src/os/windows/win32.cpp [hunk 6]
```diff
@@ -544,10 +544,10 @@ std::string FS2OTTD(const std::wstring &name)
 	int name_len = (name.length() >= INT_MAX) ? INT_MAX : (int)name.length();
 	int len = WideCharToMultiByte(CP_UTF8, 0, name.c_str(), name_len, nullptr, 0, nullptr, nullptr);
 	if (len <= 0) return std::string();
-	char *utf8_buf = AllocaM(char, len + 1);
+	std::string utf8_buf(len + 1, '\0');
```
Comment (glx22) [inline src/os/windows/win32.cpp]:
```suggestion
	std::string utf8_buf(len, '\0');
```
WideCharToMultiByte() returned length includes terminating null character.

The old code used +1 for extra safety but cropped the result on return.

File: src/os/windows/win32.cpp [hunk 7]
```diff
@@ -562,10 +562,10 @@ std::wstring OTTD2FS(const std::string &name)
 	int name_len = (name.length() >= INT_MAX) ? INT_MAX : (int)name.length();
 	int len = MultiByteToWideChar(CP_UTF8, 0, name.c_str(), name_len, nullptr, 0);
 	if (len <= 0) return std::wstring();
-	wchar_t *system_buf = AllocaM(wchar_t, len + 1);
+	std::wstring system_buf(len + 1, L'\0');
```
Comment (glx22) [inline src/os/windows/win32.cpp]:
```suggestion
	std::wstring system_buf(len, L'\0');
```
MultiByteToWideChar() returned length includes terminating null character.

The old code used +1 for extra safety but cropped the result on return.

File: src/os/windows/win32.cpp [hunk 8]
```diff
@@ -669,13 +669,13 @@ int OTTDStringCompare(const char *s1, const char *s2)
 		int len_s2 = MultiByteToWideChar(CP_UTF8, 0, s2, -1, nullptr, 0);
 
 		if (len_s1 != 0 && len_s2 != 0) {
-			LPWSTR str_s1 = AllocaM(WCHAR, len_s1);
-			LPWSTR str_s2 = AllocaM(WCHAR, len_s2);
+			std::wstring str_s1(len_s1 + 1, L'\0');
+			std::wstring str_s2(len_s2 + 1, L'\0');
```
Comment (glx22) [inline src/os/windows/win32.cpp]:
```suggestion
			std::wstring str_s1(len_s1, L'\0');
			std::wstring str_s2(len_s2, L'\0');
```
MultiByteToWideChar() returned length includes terminating null character, so not using +1 was fine in old code.

**Review by glx22 — COMMENTED** [review]
Ok I found why regression fails on windows, it couldn't find .obg and friends.
It's because FS2OTTD() returned string had an extra '\0', making the string 1 char longer and causing `if (d_name == "." || d_name == "..") continue;` to compare ".\0" with "." or ".." which returns false in both cases.

**Review by rubidium42 — APPROVED** [review]

**Pull request merged**